cloud_utils/cache/utils.py

```diff
@@ -93,42 +90,50 @@ def auto_updating_cache(
     save_local: bool,
     bucket_name: str,
     should_update: Callable[[Optional[datetime.timedelta]], bool],
+    identifier_func: Callable[[Any], str],
+    path: str,
 ) -> Callable:
-    cache_file = _get_cache_filename(cache_file_name, factory)
+    cache_file = _get_cache_filename(cache_file_name, path)
     extra_fields = {
         "filename": os.path.basename(factory.__code__.co_filename),
         "lineno": factory.__code__.co_firstlineno,
     }
-    identifier = gamla.function_to_uid(factory)
-    return gamla.compose_left(
-        gamla.just(cache_file),
-        file_store.open_file("r"),
-        json.load,
-        gamla.side_effect(
-            gamla.compose_left(
-                _time_since_last_updated(identifier),
-                _total_hours_since_update,
-                lambda hours_since_last_update: f"Loading cache for [{identifier}]. Last updated {hours_since_last_update} hours ago.",
-                logging.info,
+
+    async def inner(*args):
+        identifier = identifier_func(*args)
+
+        return await gamla.pipe(
+            cache_file,
+            file_store.open_file("r"),
+            json.load,
+            gamla.side_effect(
+                gamla.compose_left(
+                    _time_since_last_updated(identifier),
+                    _total_hours_since_update,
+                    lambda hours_since_last_update: f"Loading cache for [{identifier}]. Last updated {hours_since_last_update} hours ago.",
+                    logging.info,
+                ),
             ),
-        ),
-        gamla.ternary(
-            gamla.compose_left(_time_since_last_updated(identifier), should_update),
-            gamla.compose_left(
-                gamla.ignore_input(factory),
-                file_store.save_to_bucket_return_hash(save_local, bucket_name),
-                gamla.side_effect(
-                    _write_to_cache_file(
-                        cache_file,
-                        identifier,
-                        extra_fields,
+            gamla.ternary(
+                gamla.compose_left(_time_since_last_updated(identifier), should_update),
+                gamla.compose_left(
+                    gamla.just(factory),
+                    gamla.apply_async(*args),
```

Comment:
factory might be sync. so we need to check and use the correct apply, consider adding to gamla (generic apply)